tsdb/head.go

```diff
@@ -1218,6 +1256,78 @@ func (a *headAppender) AddFast(ref uint64, t int64, v float64) error {
 	return nil
 }
 
+func (a *headAppender) AddExemplar(lset labels.Labels, e exemplar.Exemplar) error {
+	// this should be t && e.Ts < a.minValidTime?
+	if e.Ts < a.minValidTime {
+		// exemplar out of bounds?
+		return storage.ErrOutOfBounds
+	}
+
+	// Ensure no empty labels have gotten through.
+	lset = lset.WithoutEmpty()
+
+	if len(lset) == 0 {
+		return errors.Wrap(ErrInvalidExemplar, "empty labelset")
+	}
+
+	if l, dup := lset.HasDuplicateLabelNames(); dup {
+		return errors.Wrap(ErrInvalidExemplar, fmt.Sprintf(`label name "%s" is not unique`, l))
+	}
+
+	s, created, err := a.head.getOrCreate(lset.Hash(), lset)
+	if err != nil {
+		return err
+	}
+
+	// in theory the series should never be created as a result of an AddExemplar call, I think
+	if created {
+		a.series = append(a.series, record.RefSeries{
+			Ref:    s.ref,
+			Labels: lset,
+		})
+	}
+	return a.AddExemplarFast(s.ref, e)
+}
+
+func (a *headAppender) AddExemplarFast(ref uint64, e exemplar.Exemplar) error {
+	if e.Ts < a.minValidTime {
+		// todo metric for exemplar out of bounds? or use sample out of bounds?
+		return storage.ErrOutOfBounds
+	}
+
+	s := a.head.series.getByID(ref)
+	if s == nil {
+		return errors.Wrap(storage.ErrNotFound, "unknown series")
+	}
+
+	// todo: fix timestamp checks
+	// s.Lock()
+	// if err := s.appendable(e.Ts, e.V); err != nil {
+	// 	s.Unlock()
+	// 	if err == storage.ErrOutOfOrderSample {
+	// 		a.head.metrics.outOfOrderSamples.Inc()
+	// 	}
+	// 	return err
+	// }
+	// s.pendingCommit = true
+	// s.Unlock()
+
+	// if t < a.mint {
+	// 	a.mint = t
+	// }
+	// if t > a.maxt {
+	// 	a.maxt = t
+	// }
+
+	a.exemplars = append(a.exemplars, record.RefExemplar{
```

Comment:
We should check if the exemplar is appendable before adding here

Reply:
I think based on the conversation we had on Slack we're not going to do this until we deal with WAL records in the next PR, and we'd call the exemplar storages `AddExemplar` to decide whether to add the exemplar to the WAL. Let me know if that doesn't sound right.